In [1]:
# import necessary modules
import pandas as pd
import numpy as np 
import seaborn as sns
import matplotlib.pyplot as plt


from IPython.display import display

### Load the data & preview the data

In [2]:
data = pd.read_csv('../data/raw/Copy of Ask A Manager Salary Survey 2021 (Responses) - Form Responses 1.csv')

In [3]:
print(data.info())
data.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 28061 entries, 0 to 28060
Data columns (total 18 columns):
 #   Column                                                                                                                                                                                                                                Non-Null Count  Dtype  
---  ------                                                                                                                                                                                                                                --------------  -----  
 0   Timestamp                                                                                                                                                                                                                             28061 non-null  object 
 1   How old are you?                                                                                             

Timestamp How old are you?  What industry do you work in?  \
0  4/27/2021 11:02:10            25-34   Education (Higher Education)   
1  4/27/2021 11:02:22            25-34              Computing or Tech   
2  4/27/2021 11:02:38            25-34  Accounting, Banking & Finance   
3  4/27/2021 11:02:41            25-34                     Nonprofits   
4  4/27/2021 11:02:42            25-34  Accounting, Banking & Finance   

                                  Job title  \
0        Research and Instruction Librarian   
1  Change & Internal Communications Manager   
2                      Marketing Specialist   
3                           Program Manager   
4                        Accounting Manager   

  If your job title needs additional context, please clarify here:  \
0                                                NaN                 
1                                                NaN                 
2                                                NaN                 
3                                                NaN                 
4                                                NaN                 

  What is your annual salary? (You'll indicate the currency in a later question. If you are part-time or hourly, please enter an annualized equivalent -- what you would earn if you worked the job 40 hours a week, 52 weeks a year.)  \
0                                             55,000                                                                                                                                                                                     
1                                             54,600                                                                                                                                                                                     
2                                             34,000                                                                                                                                                                                     
3                                             62,000                                                                                                                                                                                     
4                                             60,000                                                                                                                                                                                     

   How much additional monetary compensation do you get, if any (for example, bonuses or overtime in an average year)? Please only include monetary compensation here, not the value of benefits.  \
0                                                0.0                                                                                                                                                
1                                             4000.0                                                                                                                                                
2                                                NaN                                                                                                                                                
3                                             3000.0                                                                                                                                                
4                                             7000.0                                                                                                                                                

  Please indicate the currency  \
0                          USD   
1                          GBP   
2                          USD   
3                          USD   
4                          USD   

  If "Other," please indicate the currency here:   \
0                      

### Cleaning data:
- standardize column names
- Remove unnecessary columns
- convert columns to the appropriate data type
- handle missing values

In [29]:
# Make a copy of the original data
df = data.copy()

In [30]:
df.columns

Index(['Timestamp', 'How old are you?', 'What industry do you work in?',
       'Job title',
       'If your job title needs additional context, please clarify here:',
       'What is your annual salary? (You'll indicate the currency in a later question. If you are part-time or hourly, please enter an annualized equivalent -- what you would earn if you worked the job 40 hours a week, 52 weeks a year.)',
       'How much additional monetary compensation do you get, if any (for example, bonuses or overtime in an average year)? Please only include monetary compensation here, not the value of benefits.',
       'Please indicate the currency',
       'If "Other," please indicate the currency here: ',
       'If your income needs additional context, please provide it here:',
       'What country do you work in?',
       'If you're in the U.S., what state do you work in?',
       'What city do you work in?',
       'How many years of professional work experience do you have overall?',
       

**Standardize column names**

In [31]:
column_names = {
    'Timestamp': 'timestamp',
    'How old are you?': 'age_range',
    'What industry do you work in?': 'industry',
    'Job title': 'job_title',
    'If your job title needs additional context, please clarify here:': 'job_title_context',
    "What is your annual salary? (You'll indicate the currency in a later question. If you are part-time or hourly, please enter an annualized equivalent -- what you would earn if you worked the job 40 hours a week, 52 weeks a year.)":'salary(annualized)',
    'How much additional monetary compensation do you get, if any (for example, bonuses or overtime in an average year)? Please only include monetary compensation here, not the value of benefits.': 'additional_compensation',
    'Please indicate the currency':'currency',
    'If "Other," please indicate the currency here: ': 'other_currency',
    'If your income needs additional context, please provide it here:': 'income_context',
    'What country do you work in?': 'country',
    "If you're in the U.S., what state do you work in?":'state(US)',
    'What city do you work in?': 'city(US)',
    'How many years of professional work experience do you have overall?': 'years_of_experience',
    'How many years of professional work experience do you have in your field?': 'years_of_experience_in_field',
    'What is your highest level of education completed?': 'level_of_education',
    'What is your gender?': 'gender',
    'What is your race? (Choose all that apply.)': 'race'
}
df.columns = df.columns.map(column_names)

In [32]:
df.head()

timestamp age_range                       industry  \
0  4/27/2021 11:02:10     25-34   Education (Higher Education)   
1  4/27/2021 11:02:22     25-34              Computing or Tech   
2  4/27/2021 11:02:38     25-34  Accounting, Banking & Finance   
3  4/27/2021 11:02:41     25-34                     Nonprofits   
4  4/27/2021 11:02:42     25-34  Accounting, Banking & Finance   

                                  job_title job_title_context  \
0        Research and Instruction Librarian               NaN   
1  Change & Internal Communications Manager               NaN   
2                      Marketing Specialist               NaN   
3                           Program Manager               NaN   
4                        Accounting Manager               NaN   

  salary(annualized)  additional_compensation currency other_currency  \
0             55,000                      0.0      USD            NaN   
1             54,600                   4000.0      GBP            NaN   
2             34,000                      NaN      USD            NaN   
3             62,000                   3000.0      USD            NaN   
4             60,000                   7000.0      USD            NaN   

  income_context         country       state(US)     city(US)  \
0            NaN   United States   Massachusetts       Boston   
1            NaN  United Kingdom             NaN    Cambridge   
2            NaN              US       Tennessee  Chattanooga   
3            NaN             USA       Wisconsin    Milwaukee   
4            NaN              US  South Carolina   Greenville   

  years_of_experience years_of_experience_in_field level_of_education  \
0           5-7 years                    5-7 years    Master's degree   
1        8 - 10 years                    5-7 years     College degree   
2         2 - 4 years                  2 - 4 years     College degree   
3        8 - 10 years                    5-7 years     College degree   
4        8 - 10 years                    5-7 years     College degree   

       gender   race  
0       Woman  White  
1  Non-binary  White  
2       Woman  White  
3       Woman  White  
4       Woman  White

**Dropping Unnecessary Columns Or Columns with High Missing Values**

In [33]:
# drop columns with high missing values
(df.isnull().mean() *100).sort_values(ascending=False)

other_currency                  99.269449
income_context                  89.159331
job_title_context               74.120666
additional_compensation         25.996935
state(US)                       17.896725
level_of_education               0.791134
race                             0.630769
gender                           0.609387
city(US)                         0.292221
industry                         0.263711
job_title                        0.003564
timestamp                        0.000000
age_range                        0.000000
salary(annualized)               0.000000
currency                         0.000000
country                          0.000000
years_of_experience              0.000000
years_of_experience_in_field     0.000000
dtype: float64

'income_context', 'other_currency', 'job_title_context' all have more than 70% missing values, so we can drop them. But before doing so, let's check if there are any meaningful information in these columns.

In [34]:
print('Number of Non-null values in the Other_currency Column: ',len(df[~df['other_currency'].isnull()]))
print('Number of unique values in the Other_currency Column: ',df[~df['other_currency'].isnull()]['other_currency'].nunique())
df[~df['other_currency'].isnull()].iloc[:,:10].sample(20)

Number of Non-null values in the Other_currency Column:  205
Number of unique values in the Other_currency Column:  121


timestamp age_range  \
27897  11/10/2022 11:49:07     25-34   
752     4/27/2021 11:14:29     45-54   
11734    4/28/2021 4:27:20     45-54   
13218   4/28/2021 12:14:42     35-44   
27923    1/8/2023 12:16:15     18-24   
10782   4/27/2021 22:15:04     35-44   
18498    4/29/2021 0:33:04     35-44   
5729    4/27/2021 13:21:37     35-44   
8004    4/27/2021 15:28:25     35-44   
24949    5/5/2021 23:02:24     25-34   
6969    4/27/2021 14:18:47     25-34   
11668    4/28/2021 3:26:11     45-54   
9384    4/27/2021 17:50:21     25-34   
25613     5/7/2021 4:18:17     45-54   
25294    5/6/2021 12:03:34     35-44   
9921    4/27/2021 19:23:18     35-44   
1001    4/27/2021 11:18:27     35-44   
3124    4/27/2021 11:59:26     25-34   
7739    4/27/2021 15:09:39     45-54   
9845    4/27/2021 19:05:21     25-34   

                                        industry                    job_title  \
27897              Accounting, Banking & Finance      Manager, Risk Reporting   
752    Public Library (technically City Govt.?)      Administrative Librarian   
11734                     Transport or Logistics                   HR Manager   
13218                                Health care        Operational Services    
27923                          Computing or Tech           Computer Scientist   
10782              Education (Primary/Secondary)     Secondary School Teacher   
18498                            Media & Digital         TOEIC Test Developer   
5729        Government and Public Administration       Head of administration   
8004                           Computing or Tech  Senior Executive Assistant    
24949                                Translation         Freelance translator   
6969                           Computing or Tech           Software Developer   
11668                          Computing or Tech              Project Manager   
9384                Engineering or Manufacturing               Senior Analyst   
25613              Education (Primary/Secondary)               Vice Principal   
25294                          Computing or Tech    Senior software developer   
9921               Education (Primary/Secondary)         Executive Assistant    
1001                           Recruitment or HR  HR Manager/Accounts Payable   
3124                                       Sales            Territory Manager   
7739                                 Health care   Board certified physician    
9845                           Computing or Tech            Software Engineer   

                                       job_title_context salary(annualized)  \
27897                                                NaN             110000   
752    I run a branch library and am responsible for ...             76,302   
11734                                                NaN            104,917   
13218                                       Kitchen Hand             83,000   
27923                                                NaN              10000   
10782                                                NaN             90,000   
18498      Only for test-prep books, not the actual test         38,000,000   
5729                                                 NaN            813,000   
8004                                                 NaN            115,000   
24949                                                NaN             230000   
6969                                                 NaN            288,000   
11668                                                NaN            126,000   
9384                                                 NaN            108,000   
25613                                                NaN             640000   
25294                                                NaN             221000   
9921   EA to the Principal and Leadership Team + HR A...            103,516   
1001                                                 NaN             53,500   
3124              Software sales for automotive indu

There are 121 out of 205 unique values in the 'other_currency' column, with makes it a lot of unique values to deal with and the benefit of dealing with each one is minimal. So the current approach at the moment would be to first drop all the rows in the dataframe that have values in the other_currency column because they can skew the analysis of the data then drop the column itself after.

The other columns with high missing values (income_context, job_title_context) can be dropped without losing any relevant information.

In [35]:
# Remove rows where 'other_currency' column has non-null values.
df = df[df['other_currency'].isnull()]
df.shape

(27856, 18)

In [36]:
# Drop unnecessary columns.
df = df.drop(columns=['income_context', 'other_currency', 'job_title_context'],axis=1)

In [37]:
df.head()

timestamp age_range                       industry  \
0  4/27/2021 11:02:10     25-34   Education (Higher Education)   
1  4/27/2021 11:02:22     25-34              Computing or Tech   
2  4/27/2021 11:02:38     25-34  Accounting, Banking & Finance   
3  4/27/2021 11:02:41     25-34                     Nonprofits   
4  4/27/2021 11:02:42     25-34  Accounting, Banking & Finance   

                                  job_title salary(annualized)  \
0        Research and Instruction Librarian             55,000   
1  Change & Internal Communications Manager             54,600   
2                      Marketing Specialist             34,000   
3                           Program Manager             62,000   
4                        Accounting Manager             60,000   

   additional_compensation currency         country       state(US)  \
0                      0.0      USD   United States   Massachusetts   
1                   4000.0      GBP  United Kingdom             NaN   
2                      NaN      USD              US       Tennessee   
3                   3000.0      USD             USA       Wisconsin   
4                   7000.0      USD              US  South Carolina   

      city(US) years_of_experience years_of_experience_in_field  \
0       Boston           5-7 years                    5-7 years   
1    Cambridge        8 - 10 years                    5-7 years   
2  Chattanooga         2 - 4 years                  2 - 4 years   
3    Milwaukee        8 - 10 years                    5-7 years   
4   Greenville        8 - 10 years                    5-7 years   

  level_of_education      gender   race  
0    Master's degree       Woman  White  
1     College degree  Non-binary  White  
2     College degree       Woman  White  
3     College degree       Woman  White  
4     College degree       Woman  White

**Cleaning other columns with missing values**

In [38]:
(df.isnull().mean() * 100).sort_values(ascending=False)

additional_compensation         25.969271
state(US)                       17.428920
level_of_education               0.771827
race                             0.621051
gender                           0.610281
city(US)                         0.294371
industry                         0.258472
job_title                        0.003590
salary(annualized)               0.000000
timestamp                        0.000000
age_range                        0.000000
currency                         0.000000
country                          0.000000
years_of_experience              0.000000
years_of_experience_in_field     0.000000
dtype: float64

- Additional Compensation Column

In [39]:
# fill missing values in 'additional_compensation' column with 0
df['additional_compensation'] = df['additional_compensation'].fillna(0)

- State(US) Column

In [40]:
# Viewing the pattern of the state(US) column to see if United States has have corresponding states and 
# non United States entries have null values.
df.iloc[:,6:9]

currency         country       state(US)
0          USD   United States   Massachusetts
1          GBP  United Kingdom             NaN
2          USD              US       Tennessee
3          USD             USA       Wisconsin
4          USD              US  South Carolina
...        ...             ...             ...
28056      CAD          Canada             NaN
28057      USD  United States         Missouri
28058      USD             USA         Georgia
28059      USD         Myanmar        Colorado
28060      USD           Burma             NaN

[27856 rows x 3 columns]

In [41]:
# View rows with missing values in 'State(US)' column to confirm the pattern.
df[df['state(US)'].isnull()].iloc[:,6:10]

currency          country state(US)    city(US)
1          GBP   United Kingdom       NaN   Cambridge
10         USD    United States       NaN  Boston, MA
14         CAD           Canada       NaN      Remote
15         GBP  United Kingdom        NaN     Lincoln
21         USD    United States       NaN     Atlanta
...        ...              ...       ...         ...
28050      EUR            Spain       NaN   Barcelona
28053      CAD           Taiwan       NaN      Taipei
28055      CAD           Canada       NaN     Toronto
28056      CAD           Canada       NaN   Kitchener
28060      USD            Burma       NaN      Yangon

[4855 rows x 4 columns]

It appears that there are some rows where the 'country' column is 'United States' but the 'State(US)' column is missing.

Since there are multiple entries where the in the country column United States is written in many different ways, we will convert all entries to 'United States' to make it consistent first before dealing with the missing state entries.

In [42]:
df['country'].unique()

array(['United States', 'United Kingdom', 'US', 'USA', 'Canada',
       'United Kingdom ', 'usa', 'UK', 'Scotland ', 'U.S.',
       'United States ', 'The Netherlands', 'Australia ', 'Spain', 'us',
       'Usa', 'England', 'finland', 'United States of America', 'France',
       'United states', 'Scotland', 'USA ', 'United states ', 'Germany',
       'UK ', 'united states', 'Ireland', 'Australia', 'Uk',
       'United States of America ', 'U.S. ', 'canada', 'Canada ', 'U.S>',
       'ISA', 'Great Britain ', 'US ', 'United State', 'U.S.A', 'Denmark',
       'U.S.A.', 'America', 'Netherlands', 'netherlands', 'England ',
       'united states of america', 'Ireland ', 'Switzerland',
       'Netherlands ', 'Bermuda', 'Us', 'The United States',
       'United State of America', 'Germany ', 'Mexico ', 'United Stated',
       'South Africa ', 'Belgium', 'Northern Ireland', 'u.s.',
       'South Africa', 'UNITED STATES', 'united States', 'Sweden',
       'Hong Kong', 'Sri lanka', 'Contracts', 'U

Starting with USA, we can see it was entered in many ways. We will convert them all to 'United States'.

- Convert all USA entries in 'country' column to 'United States'.

In [43]:
# First make all the entries in 'country' column lowercase.
df['country'] = df['country'].str.lower().str.strip()

In [44]:
# show all unique country entries that start with the letter 'u'
df[df['country'].str.startswith('u')]['country'].unique()

array(['united states', 'united kingdom', 'us', 'usa', 'uk', 'u.s.',
       'united states of america', 'u.s>', 'united state', 'u.s.a',
       'u.s.a.', 'united state of america', 'united stated',
       'usa-- virgin islands', 'united statws', 'u.s', 'unites states',
       'u. s.', 'united sates', 'united states of american',
       'uniited states', 'united kingdom (england)',
       'united sates of america',
       'united states (i work from home and my clients are all over the us/canada/pr',
       'unted states', 'united statesp', 'united stattes',
       'united statea', 'united kingdom.', 'united statees',
       'uniyed states', 'uniyes states', 'united states of americas',
       'u.a.', 'us of a', 'united arab emirates', 'u.k.', 'u.sa',
       'united kindom', 'united status', 'uxz', 'uss', 'uniteed states',
       'united stares', 'uk (northern ireland)', 'uk for u.s. company',
       'unite states', 'united kingdomk', 'unitedstates',
       'u.k. (northern england)', 'u

In [48]:
# Get all the usa varitions into list
usa_list = ['united states', 'united kingdom', 'us', 'usa', 'u.s.',
       'united states of america', 'u.s>', 'united state', 'u.s.a',
       'u.s.a.', 'united state of america', 'united stated',
       'usa-- virgin islands', 'united statws', 'u.s', 'unites states',
       'u. s.', 'united sates', 'united states of american',
       'uniited states', 'united sates of america',
       'united states (i work from home and my clients are all over the us/canada/pr',
       'unted states', 'united statesp', 'united stattes',
       'united statea', 'united statees',
       'uniyed states', 'uniyes states', 'united states of americas',
       'u.a.', 'us of a', 'u.sa',
       'united status', 'uniteed states',
       'united stares', 'unite states', 'unitedstates',
       'united statew', 'united statues', 'untied states',
       'usa (company is based in a us territory, i work remote)',
       'usab', 'unitied states',
       'united sttes', 'uniter statez', 'u. s',
       'usa tomorrow', 'united stateds',
       'usat', 'unitef stated', 'ua', 'usaa',
       'united y', 'united statss', 'united  states',
       'united states is america','america','isa','the united states',
       'san francisco', '🇺🇸','california']

# replace all USA entries in 'country' column with 'United States' using the list usa_list
df['country'] = df['country'].apply(lambda x: 'united states' if x in usa_list else x)

In [49]:
df['country'].unique()

array(['united states', 'canada', 'uk', 'scotland', 'the netherlands',
       'australia', 'spain', 'england', 'finland', 'france', 'germany',
       'ireland', 'great britain', 'denmark', 'netherlands',
       'switzerland', 'bermuda', 'mexico', 'south africa', 'belgium',
       'northern ireland', 'sweden', 'hong kong', 'sri lanka',
       'contracts', 'england/uk',
       "we don't get raises, we get quarterly bonuses, but they periodically asses income in the area you work, so i got a raise because a 3rd party assessment showed i was paid too little for the area we were located",
       'england, uk.', 'greece', 'japan', 'britain', 'austria',
       'canada, ottawa, ontario', 'global', 'united kingdom (england)',
       'worldwide (based in us but short term trips aroudn the world)',
       'canadw', 'luxembourg', 'united kingdom.', 'new zealand',
       'cayman islands', 'can',
       'i am located in canada but i work for a company in the us',
       'latvia', 'puerto rico', 'rwa

In [50]:
# Get the uk entries in 'country' column to 'United Kingdom'
df[df['country'].str.startswith('u')]['country'].unique()

array(['united states', 'uk', 'united kingdom (england)',
       'united kingdom.', 'united arab emirates', 'u.k.', 'united kindom',
       'uxz', 'uss', 'uk (northern ireland)', 'uk for u.s. company',
       'united kingdomk', 'u.k. (northern england)', 'u.k', 'uae',
       'uk (england)', 'uk, remote', 'unites kingdom', 'uruguay',
       'us govt employee overseas, country withheld', 'uganda', 'ukraine',
       'united states- puerto rico',
       'uk, but for globally fully remote company', 'usd',
       "usa, but for foreign gov't"], dtype=object)

In [53]:
# Get all the usa varitions into list
uk_list = ['uk', 'united kingdom (england)',
       'united kingdom.', 'u.k.', 'united kindom',
       'uk (northern ireland)', 'uk for u.s. company',
       'united kingdomk', 'u.k. (northern england)', 'u.k',
       'uk (england)', 'uk, remote', 'unites kingdom',
       'uk, but for globally fully remote company','england','england/uk',
       'england, uk.','britain','england, uk','wales (united kingdom)', 'england, gb',
       'england, united kingdom','englang','scotland, uk','wales, uk',
       'wales (uk)','northern ireland, united kingdom','london']

# replace all USA entries in 'country' column with 'United States' using the list usa_list
df['country'] = df['country'].apply(lambda x: 'united kingdom' if x in uk_list else x)

In [54]:
df['country'].unique()

array(['united states', 'canada', 'united kingdom', 'scotland',
       'the netherlands', 'australia', 'spain', 'finland', 'france',
       'germany', 'ireland', 'great britain', 'denmark', 'netherlands',
       'switzerland', 'bermuda', 'mexico', 'south africa', 'belgium',
       'northern ireland', 'sweden', 'hong kong', 'sri lanka',
       'contracts',
       "we don't get raises, we get quarterly bonuses, but they periodically asses income in the area you work, so i got a raise because a 3rd party assessment showed i was paid too little for the area we were located",
       'greece', 'japan', 'austria', 'canada, ottawa, ontario', 'global',
       'worldwide (based in us but short term trips aroudn the world)',
       'canadw', 'luxembourg', 'new zealand', 'cayman islands', 'can',
       'i am located in canada but i work for a company in the us',
       'latvia', 'puerto rico', 'rwanda', 'united arab emirates',
       'romania', 'currently finance', 'serbia', 'russia', 'uxz', 'cand

In [31]:
# make a dict to map all USA entries to 'United States'.
us_to_usa_map = {
    'usa': 'united states',
    'us': 'united states',
    'united states': 'united states',
    'united states of america': 'united states',
    'u.s.': 'united states',
    'united States': 'united states',
    'united states ': 'united states',
    'united stated': 'united states',
    'usa': 'united states',
    'uSA': 'united states',
    'uS': 'united states'
}

# replace all USA entries in 'country' column with 'United States'.
df['country'] = df['country'].replace(us_to_usa_map)

In [32]:
df['country'].unique()

array(['United States', 'United Kingdom', 'US', 'USA', 'Canada',
       'United Kingdom ', 'united states', 'UK', 'Scotland ', 'U.S.',
       'United States ', 'The Netherlands', 'Australia ', 'Spain', 'Usa',
       'England', 'finland', 'United States of America', 'France',
       'United states', 'Scotland', 'USA ', 'United states ', 'Germany',
       'UK ', 'Ireland', 'Australia', 'Uk', 'United States of America ',
       'U.S. ', 'canada', 'Canada ', 'U.S>', 'ISA', 'Great Britain ',
       'US ', 'United State', 'U.S.A', 'Denmark', 'U.S.A.', 'America',
       'Netherlands', 'netherlands', 'England ', 'Ireland ',
       'Switzerland', 'Netherlands ', 'Bermuda', 'Us',
       'The United States', 'United State of America', 'Germany ',
       'Mexico ', 'United Stated', 'South Africa ', 'Belgium',
       'Northern Ireland', 'South Africa', 'UNITED STATES', 'Sweden',
       'Hong Kong', 'Sri lanka', 'Contracts', 'USA-- Virgin Islands',
       'United Statws', 'England/UK', 'U.S',
      